src/Cache/FileHandler.php

```diff
@@ -60,17 +60,39 @@ public function write(CacheInterface $cache)
     {
         $content = $cache->toJson();
 
+        if (file_exists($this->file)) {
+            if (is_dir($this->file)) {
+                throw new IOException(
+                    sprintf('Cannot write cache file "%s" as the location exists as directory.', $this->file),
+                    0,
+                    null,
+                    $this->file
+                );
+            }
+
+            if (!is_writable($this->file)) {
+                throw new IOException(
+                    sprintf('Cannot write to file "%s" as it is not writable.', $this->file),
+                    0,
+                    null,
+                    $this->file
+                );
+            }
+        }
+
         $bytesWritten = @file_put_contents($this->file, $content, LOCK_EX);
 
         if (false === $bytesWritten) {
             $error = error_get_last();
 
             throw new IOException(
-                sprintf('Failed to write file "%s", "%s".', $this->file, $error ? $error['message'] : 'no reason available'),
+                sprintf('Failed to write file "%s", "%s".', $this->file, isset($error['message']) ? $error['message'] : 'no reason available'),
                 0,
                 null,
                 $this->file
             );
         }
+
+        @chmod($this->file, 0666);
```

Comment:
I raised 2 concerns here, one about umask, which you answered, and 2nd one:
> why setting the permission of file after we already write to it ?

Reply:
>why setting the permission of file after we already write to it ?

If I run the fixer using `sudo` on Ubuntu box with default `umask` the cache file is created without write permissions for users. Like reported in https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3122
If you run the fixer again without `sudo` it will fail to write to the previous created cache file because of lack of permission to do so.

Reply:
it needs to be done only after file creation, not after every single writing to file.

Reply:
:+1: done